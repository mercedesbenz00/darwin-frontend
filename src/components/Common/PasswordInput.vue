<template>
  <input-field
    v-model="password"
    class="password-input"
    name="password"
    type="password"
    label="Password"
    theme="light"
    :error="error"
    hint="Password needs to have at least 8 characters"
    required="required"
    :autocomplete="autocomplete"
  >
    <div
      v-if="password && passwordIssues.length === 0"
      slot="modifier"
      class="password-input__strong"
    >
      <div class="password-input__strong__dot" />
      <label>Strong Password</label>
    </div>
    <div
      v-if="password && passwordIssues.length > 0"
      slot="modifier"
      class="password-input__weak"
    >
      <label>{{ passwordIssues[0] }}</label>
    </div>
  </input-field>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator'

import InputField from '@/components/Common/InputField/V1/InputField.vue'
import { validatePassword } from '@/utils'

@Component({
  name: 'password-input',
  components: { InputField }
})
export default class PasswordInput extends Vue {
  @Prop({ required: true, type: String })
  value!: string

  @Prop({ required: false, type: String, default: 'password' })
  autocomplete!: string

  @Prop({ required: false, type: String })
  error?: string

  passwordIssues: Array<string> = []

  get password () {
    return this.value
  }

  set password (val: string) {
    this.$emit('input', val)
  }

  @Watch('password')
  onPassword () {
    this.passwordIssues = validatePassword(this.password)
  }
}
</script>

<style lang="scss" scoped>
.password-input__strong {
  position: absolute;
  @include row--center;
  @include typography(md, default, bold);
  right: 5px;
  text-align: right;
  white-space: nowrap;
  color: $colorPrimaryLight;
}

.password-input__strong__dot {
  position: absolute;
  left: -10px;
  width: 0;
  height: 0;
  border-radius: 3px;
  border: 3px solid $colorPrimaryLight;
}

.password-input__weak {
  position: absolute;
  @include row--center;
  @include typography(md, default, bold);
  right: 5px;
  text-align: right;
  color: $colorYellow;
  white-space: nowrap;
}
</style>
