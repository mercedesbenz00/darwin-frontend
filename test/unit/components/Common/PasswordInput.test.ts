import { createLocalVue, shallowMount, Stubs } from '@vue/test-utils'

import PasswordInput from '@/components/Common/PasswordInput.vue'

const localVue = createLocalVue()

let propsData: {
  value: string
  autocomplete?: string
  error?: string
}

const stubs: Stubs = {
  'input-field': {
    template: `
      <div class="input-field-stub">
        <slot name="modifier" />
      </div>
    `
  }
}

it('matches snapshot', () => {
  propsData = { value: '' }
  const wrapper = shallowMount(PasswordInput, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with strong password', () => {
  propsData = { value: 'Password1' }
  const wrapper = shallowMount(PasswordInput, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with invalid password', async () => {
  propsData = { value: '' }
  const wrapper = shallowMount(PasswordInput, { localVue, propsData, stubs })
  await wrapper.setProps({ value: 'invalid-password' })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with error', () => {
  propsData = {
    error: 'Error',
    value: 'Password1'
  }
  const wrapper = shallowMount(PasswordInput, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})
