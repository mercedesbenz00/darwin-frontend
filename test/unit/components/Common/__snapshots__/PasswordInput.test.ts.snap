// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `<div class="input-field-stub password-input" name="password" type="password" label="Password" theme="light" hint="Password needs to have at least 8 characters" required="required" autocomplete="password" value=""></div>`;

exports[`matches snapshot with error 1`] = `
<div class="input-field-stub password-input" name="password" type="password" label="Password" theme="light" error="Error" hint="Password needs to have at least 8 characters" required="required" autocomplete="password" value="Password1">
  <div class="password-input__strong">
    <div class="password-input__strong__dot"></div> <label>Strong Password</label>
  </div>
</div>
`;

exports[`matches snapshot with invalid password 1`] = `
<div class="input-field-stub password-input" name="password" type="password" label="Password" theme="light" hint="Password needs to have at least 8 characters" required="required" autocomplete="password" value="invalid-password">
  <div class="password-input__weak"><label>At least one uppercase letter</label></div>
</div>
`;

exports[`matches snapshot with strong password 1`] = `
<div class="input-field-stub password-input" name="password" type="password" label="Password" theme="light" hint="Password needs to have at least 8 characters" required="required" autocomplete="password" value="Password1">
  <div class="password-input__strong">
    <div class="password-input__strong__dot"></div> <label>Strong Password</label>
  </div>
</div>
`;
